### PR TITLE
Strip some control characters from links

### LINKED ISSUE #30: uri_schemes => [undef] opens door for javascript: scheme
I tried the following:
```perl
my $r = new HTML::Restrict(
    rules   => { a => [ 'href' ] },
    uri_schemes => [ undef ],
);

for my $byte (1 .. 8, 14 .. 31) {
    say $r->process(qq(<a href="&#$byte;javascript:alert(1);">click me</a>));
}
```
None of these links are filtered, but if I open the output in a browser (tested in Firefox 60 and Chrome 70), Javascript gets executed after click on any link.

Same behaviour can be observed with hexadecimal entities (`sprintf('&#x%x;', $byte)`) or with `chr($byte)` instead of entities.

Maybe this is an issue of browsers behaving incorrectly, or of the `URI` module, but I guess it's better to let you know.

### PR DESCRIPTION
**Pull request by oalders**
Fixes #30

### CODE REVIEW

File: lib/HTML/Restrict.pm [hunk 1]
```diff
@@ -150,16 +150,34 @@ sub _build_parser {
...
+                        # characters.  Browsers will helpfully ignore them,
+                        # meaning that these characters can be used to defeat
+                        # HTML::Restrict when used as leading characters in a
+                        # link.  In our case we will strip them all regardless
+                        # of where they are in the URL. See
+                        # https://github.com/oalders/html-restrict/issues/30
+
+                        if ($link) {
+                            $link =~ s/[\001-\010]/ /g;    # decimal 1..8
+                            $link =~ s/[\016-\037]/ /g;    # decimal 14..31
```
Comment (oschwald) [inline lib/HTML/Restrict.pm]: Is there a reason why you picked this set of characters? Does the null byte or a vertical tab not have the same issues? What about Unicode control characters? I wonder if something like `[[:cntrl:]]` or `[^[:print:]]` would be preferable.

Comment (oalders) [reply lib/HTML/Restrict.pm]: So, the null byte doesn't suffer from this problem.  (At least not in Chrome). Assuming `<a href="&#0;javascript:alert(1);">0</a><br>` is the correct way to test it.

However, I did some more reading at https://infra.spec.whatwg.org/#c0-control and it looks like "U+0000 NULL to U+001F INFORMATION SEPARATOR ONE" inclusive are not valid in a URL, so there's no reason not to include the whole set.  I'll make some changes.

There is no mention of C1 control codes at the spec I've been working from: https://url.spec.whatwg.org/

Comment (oschwald) [reply lib/HTML/Restrict.pm]: It doesn't work for the null byte, but it does work for the vertical tab and other control characters that are currently being excluded.

Comment (oalders) [reply lib/HTML/Restrict.pm]: Interesting.  I tested them on Mac Chrome, Firefox and Safari and none of the excluded characters created a dialog box.

Comment (oalders) [reply lib/HTML/Restrict.pm]: I've reworked the PR to remove the whole set.

Comment (oschwald) [reply lib/HTML/Restrict.pm]: Weird. I tested`<a href="&#0011;javascript:alert(1);">0</a>` and `<a href="&#xB;javascript:alert(1);">0</a>` on both Firefox and Chrome on Ubuntu and saw the dialog with both.

File: lib/HTML/Restrict.pm [hunk 2]
```diff
@@ -151,16 +151,38 @@ sub _build_parser {
...
+                        # https://url.spec.whatwg.org/
+                        # https://infra.spec.whatwg.org/#c0-control
+
+                        if ($link) {
+                            $link =~ s/[\00-\037]/ /g
+                                ;    # decimal 0..31 C0 control chars
+                            my $url = URI->new($link)->as_string;
+
+                            # The above regex doesn't strip the null byte
+                            $url =~ s{&#0;}{}g;
```
Comment (oschwald) [inline lib/HTML/Restrict.pm]: I am not clear on why this is necessary. Also, it doesn't seem particularly effective given that `&#0000`, `&x0`, etc, are all equivalent. Maybe it isn't even worth worrying about if browsers don't seem to have the behavior on null bytes.

File: lib/HTML/Restrict.pm [hunk 3]
```diff
@@ -151,16 +151,38 @@ sub _build_parser {
 
                     foreach my $source_type ( 'href', 'src', 'cite' ) {
 
-                        if ( $attr->{$source_type} ) {
-                            my $uri = URI->new( $attr->{$source_type} );
-                            if ( defined $uri->scheme ) {
+                        my $link = $attr->{$source_type};
+
+                        # Remove unprintable ASCII control characters, which
+                        # are 1..8 and 14..31. These characters are not valid
```
Comment (oschwald) [inline lib/HTML/Restrict.pm]: This might need to be updated.

File: t/control-characters.t [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+use Test::More;
+
+use HTML::Restrict ();
+
+my $r = HTML::Restrict->new(
+    rules       => { a => ['href'] },
+    uri_schemes => [undef],
+);
+
+for my $i ( 0 .. 31 ) {
```
Comment (oschwald) [inline t/control-characters.t]: Maybe it would be worth testing the `&#x...` form?

File: lib/HTML/Restrict.pm (hunk 2 above)
Comment (oalders) [reply lib/HTML/Restrict.pm]: Good point.  For some reason the regex above isn't catching it.  It doesn't seem to cause a problem for browsers, but the spec does say that it should be removed.

File: lib/HTML/Restrict.pm [hunk 5]
```diff
@@ -151,16 +151,42 @@ sub _build_parser {
...
+                        if ($link) {
+
+                            # C0 control chars (decimal 0..31)
+                            # sort of like $link =~ s/[[:^print:]]//g
+                            $link =~ s/[\00-\037]/ /g;
+
+                            my $url = URI->new($link)->as_string;
+
+                            ## The above regex doesn't strip the null byte
+                            $url =~ s{&#x?0+;}{}g;
```
Comment (oschwald) [inline lib/HTML/Restrict.pm]: The issue here seems to be that `HTML::Parser` doesn't decode `&#0;` and friends. I think this is a better solution thank what you have:

```
diff --git a/lib/HTML/Restrict.pm b/lib/HTML/Restrict.pm
index 650dd8a..6040d7b 100644
--- a/lib/HTML/Restrict.pm
+++ b/lib/HTML/Restrict.pm
@@ -171,14 +171,9 @@ sub _build_parser {
 
                             # C0 control chars (decimal 0..31)
                             # sort of like $link =~ s/[[:^print:]]//g
-                            $link =~ s/[\00-\037]/ /g;
+                            $link =~ s/[\00-\037]|&#x?0+;/ /g;
 
-                            my $url = URI->new($link)->as_string;
-
-                            ## The above regex doesn't strip the null byte
-                            $url =~ s{&#x?0+;}{}g;
-
-                            $url = URI->new($url);
+                            my $url = URI->new($link);
                             if ( defined $url->scheme ) {
                                 delete $attr->{$source_type}
                                     if none { $_ eq $url->scheme }
```

Comment (oalders) [reply lib/HTML/Restrict.pm]: Yep, that's better.

**Pull request merged**